### PR TITLE
Allow buttons to work on monitors

### LINKED ISSUE #4: Support for rendering and interaction on monitor
right now, the flow I use is
```lua
moni.clear()
prime.button(moni, ...)
parallel.waitForAny(prime.run, function()
  while true do
    -- event loop
  end
end)
```
this works, but interactions are on the computer, not the monitor (touch events, keyboard-only components don't make sense). if I get to it, I'll pr something tmr, otherwise take this quick sample I wrote in 5 minutes (take it with a grain of salt - there's probably something I forgor) - note this is only for buttons, but the general concept stays
```lua
function PrimeUI.button(win, x, y, text, action, fgColor, bgColor, clickedColor, periphName) -- periphName = string | nil
  -- ...
  if event == "mouse_click" and periphName == nil and button == 1 and clickX >= screenX and clickX < screenX + #text + 2 and clickY == screenY then
    -- ...
  elseif event == "monitor_touch" and periphName == button and clickX >= screenX and clickX < screenX + #text + 2 and clickY == screenY
    or event == "mouse_up" and button == 1 and buttonDown then -- buttonDown is false because the mouse_click event is blocked when clicked on the screen
    -- ..
  end
end

-----------

prime.button(monitorRight, 2, 4, string.char(25), function()
  -- ...
end, nil, nil, nil, "right")) -- this could probably be stripped down by using the win var, not sure
```

### PR DESCRIPTION
**Pull request by ghost**
I'm not actually sure how well this works, I'll have to test it more.

I've also made this in around 15 minutes, so it's not tested much.
Should this use the button component or be split into a different one?
Fixes #4 (i think)
Closes #5

### CODE REVIEW

File: button.lua [hunk 1]
```diff
@@ -20,6 +21,10 @@ function PrimeUI.button(win, x, y, text, action, fgColor, bgColor, clickedColor)
     fgColor = expect(6, fgColor, "number", "nil") or colors.white
     bgColor = expect(7, bgColor, "number", "nil") or colors.gray
     clickedColor = expect(8, clickedColor, "number", "nil") or colors.lightGray
+    do
```
Comment (MCJack123) [inline button.lua]: I'd appreciate adding a few comments describing what each part does.

File: button.lua [hunk 2]
```diff
@@ -20,6 +21,10 @@ function PrimeUI.button(win, x, y, text, action, fgColor, bgColor, clickedColor)
     fgColor = expect(6, fgColor, "number", "nil") or colors.white
     bgColor = expect(7, bgColor, "number", "nil") or colors.gray
     clickedColor = expect(8, clickedColor, "number", "nil") or colors.lightGray
```
Comment (MCJack123) [inline button.lua:23]: There should be a type check for the new argument.

**Review by alessiodam — APPROVED** [review]
Works fine and doesn't affect performance.

File: button.lua [hunk 3]
```diff
@@ -28,17 +33,24 @@ function PrimeUI.button(win, x, y, text, action, fgColor, bgColor, clickedColor)
...
+            then
                 -- Initiate a click action (but don't trigger until mouse up).
                 buttonDown = true
                 -- Redraw the button with the clicked background color.
                 win.setCursorPos(x, y)
                 win.setBackgroundColor(clickedColor)
                 win.setTextColor(fgColor)
                 win.write(" " .. text .. " ")
+
+                if monitor then
```
Comment (MCJack123) [inline button.lua:56]: As above, comment about why we need a timer here.

File: button.lua [hunk 4]
```diff
@@ -47,11 +59,19 @@ function PrimeUI.button(win, x, y, text, action, fgColor, bgColor, clickedColor)
                     else action() end
                 end
                 -- Redraw the original button state.
+                win.setCursorPos(x, y)
+                win.setBackgroundColor(bgColor)
+                win.setTextColor(fgColor)
+                win.write(" " .. text .. " ")
+            elseif event == "timer" and timer == button and buttonDown then
```
Comment (MCJack123) [inline button.lua]: This should probably be moved with the `mouse_up` handler to reduce code duplication.

**Review by MCJack123 — CHANGES_REQUESTED** [review]
Whoops, I forgor to submit my review for this. Looks good, but I just need a few tiny tweaks to smoothen it out.

File: button.lua (hunk 1 above)
Comment (ghost) [reply button.lua]: I'll just rewrite it to be saner.

File: button.lua (hunk 4 above)
Comment (ghost) [reply button.lua]: I don't actually know why i did that :kekw:

File: button.lua (hunk 3 above)
Comment (ghost) [reply button.lua:56]: Done.

**Comment by ghost** [comment]
Actually now it requires the monitor argument, should I leave it like that or see if the window is a monitor and use that?

**Comment by MCJack123** [comment]
True, I would prefer if you auto-detect monitors through `pcall`ing `peripheral.getName` - that would make it less confusing for newcomers.

**Comment by viwty** [comment]
Alright, though I'll have to do that tomorrow.

**Comment by ghost** [comment]
I love github mobile bugs.
Alright, but I'll have to do it tomorrow.

**Comment by ghost** [comment]
Finally got around to doing it, also fixed a typo that broke everything (oops).

**Pull request closed without merging**